crates/circuit/src/symbol_expr.rs

```diff
@@ -710,14 +735,33 @@ impl SymbolExpr {
     }
 
     /// return hashset of all symbols this equation contains
-    pub fn symbols(&self) -> HashSet<String> {
+    pub fn symbols(&self) -> HashSet<Self> {
         match self {
-            SymbolExpr::Symbol(e) => HashSet::<String>::from([e.as_ref().clone()]),
-            SymbolExpr::Value(_) => HashSet::<String>::new(),
+            SymbolExpr::Symbol { .. } => HashSet::<Self>::from([self.clone()]),
+            SymbolExpr::Value(_) => HashSet::<Self>::new(),
             SymbolExpr::Unary { op: _, expr } => expr.symbols(),
             SymbolExpr::Binary { op: _, lhs, rhs } => {
-                let mut symbols = HashSet::<String>::new();
+                let mut symbols = HashSet::<Self>::new();
                 for s in lhs.symbols().union(&rhs.symbols()) {
+                    symbols.insert(s.clone());
+                }
+                symbols
+            }
+        }
+    }
+
+    /// return hashset of all symbols this equation contains
+    pub fn symbols_in_string(&self) -> HashSet<String> {
+        match self {
+            SymbolExpr::Symbol { name, index } => match index {
```

Comment:
It might be better to have an enum with two variants inside `SymbolExpr::Symbol(<some enum>)` to distinguish between a normal symbol and one that indexes into something.